src/Polysemy/Final.hs

```diff
@@ -208,77 +169,86 @@ liftS m = do
--- by interpreting one of the monads in terms of the other one.
---
--- This allows for the use of operations that rely on a @'LastMember' ('Lift' m)@
--- constraint, as long as @m@ can be transformed to the final monad;
--- but be warned, this breaks the implicit contract of @'LastMember' ('Lift' m)@
--- that @m@ /is/ the final monad, so depending on the final monad and operations
--- used, 'runFinalLift' may become /unsafe/.
+-- | Run a 'Sem' containing only @'Final' 'Identity'@ as a pure value.
+runFinalPure :: forall a. Sem '[Final Identity] a -> a
+runFinalPure = coerce (runFinal @Identity @a)
+{-# INLINE runFinalPure #-}
+
+------------------------------------------------------------------------------
+-- | Given natural transformations between @m1@ and @m2@, run a @'Final' m1@
+-- effect by transforming it into a @'Final' m2@ effect.
+finalToFinal :: forall m1 m2 a r
+              . (Member (Final m2) r, Functor m2)
+             => (forall x. m1 x -> m2 x)
+             -> (forall x. m2 x -> m1 x)
+             -> Sem (Final m1 ': r) a
+             -> Sem r a
+finalToFinal to from =
+  let
+    go :: Sem (Final m1 ': r) x -> Sem r x
+    go = usingSem $ \u -> case decomp u of
+      Right (Weaving (WithWeaving wav) s wv ex ins) ->
+        fmap ex $ withWeaving $ \s' wv' ins'
+          -> fmap getCompose $ to $
+                wav
+                  (Compose (s <$ s'))
+                  (from . fmap Compose . wv' . fmap (go . wv) . getCompose)
+                  (ins' . getCompose >=> ins)
+      Left g -> liftSem (hoist go g)
+    {-# INLINE go #-}
+  in
+    go
+{-# INLINE finalToFinal #-}
+
+------------------------------------------------------------------------------
+-- | Run a @'Final' ('Sem' r)@ effect if the remaining effect stack is @r@.
 --
--- For example, 'runFinalLift' is unsafe with 'Polysemy.Async.asyncToIO' if
--- the final monad is non-deterministic, or a continuation
--- monad.
-runFinalLift :: Monad m
-              => (forall x. n x -> m x)
-              -> Sem [Final m, Embed m, Embed n] a
-              -> m a
-runFinalLift nat = usingSem $ \u -> case decomp u of
+-- This is sometimes useful for interpreters that make use of
+-- 'reinterpret', 'raiseUnder', or any of their friends.
+runFinalSem :: Sem (Final (Sem r) ': r) a -> Sem r a
+runFinalSem = usingSem $ \u -> case decomp u of
   Right (Weaving (WithWeaving wav) s wv ex ins) ->
-    ex <$> wav s (runFinalLift nat . wv) ins
-  Left g -> case decomp g of
-    Right (Weaving (Embed m) s _ ex _) -> ex . (<$ s) <$> m
-    Left g' -> case extract g' of
-      Weaving (Embed n) s _ ex _ -> ex . (<$ s) <$> nat n
-{-# INLINE runFinalLift #-}
+    ex <$> wav s (runFinalSem . wv) ins
+  Left g -> liftSem (hoist runFinalSem g)
+{-# INLINE runFinalSem #-}
+
+------------------------------------------------------------------------------
+-- | Run a @'Final' m@ effect by providing an explicit lowering function.
+lowerFinal :: Member (Embed m) r
```

Comment:
How many times does `f` get called here? Is it one for each action in `Final`, or is it one time total?

Reply:
For each action in `Final`, so this inherits the work duplication issue of traditional `lower-` interpreters.

Reply:
In that case I think I'm confused. Can I write this?

```haskell
runStateIO :: Member (Final IO) r => s -> Sem (State s ': r) a -> Sem r a
runStateIO s m = do
  ref <- final $ newIORef s
  runStateInIORef ref m
```

and if so, will it allocate the ioref exactly once?

Reply:
That will allocate the `IORef` exactly once _unless_ you use `lowerFinal`, and you use `runStateIO` in the argument of `lowerFinal`.

So this will allocate the `IORef` exactly once:
```
runFinal . runStateIO s . ...
```

But say you instead do this:
```
(runFinal . embedToFinal . runStateIO s .@ lowerFinal) . ...
``` 
Then _this_ will allocate the `IORef` multiple times.

Reply:
I'll add a _Beware_ to `lowerFinal` to explain this.

Reply:
wat. Is `lowerFinal` going to go away when we gut all the other lowers and `.@`?

Reply:
You're right, it doesn't belong on Polysemy proper. I basically added it as a contingency.

I can only think of one extremely contrived use-case:

1. You want your final monad to be `m`.
2. You have a bunch of `Final n` constraints, where `n` is different from `m`.
3. You only have a natural transformation `n` to `m`, not `m` to `n`, so `finalToFinal` won't work.
4. BUT you're able to interpret the rest of your effect stack either in terms of `m` or `n`.

In this case, you can use `lowerFinal` and `runEmbedded` to make it all work out.

But honestly, yeah. `lowerFinal` can stay here. I also don't have much justification behind `runFinalSem`, so it can stay here too. It's mostly just neat that you can implement `lowerFinal` using it.